components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt

```diff
@@ -6,15 +6,34 @@ package mozilla.components.concept.engine.prompt
 
 import android.content.Context
 import android.net.Uri
-import mozilla.components.concept.storage.Login
 import mozilla.components.concept.engine.prompt.PromptRequest.Authentication.Level
 import mozilla.components.concept.engine.prompt.PromptRequest.Authentication.Method
 import mozilla.components.concept.engine.prompt.PromptRequest.TimeSelection.Type
+import mozilla.components.concept.storage.Login
 
 /**
  * Value type that represents a request for showing a native dialog for prompt web content.
  */
 sealed class PromptRequest {
+    /**
+     * Unique identifier allowing differentiating between:
+     * - [PromptRequest] subclasses.
+     * - [PromptRequest] subclass instances.
+     */
+    val uid = this.hashCode()
+
+    override fun hashCode(): Int {
+        // Salt the hashcode to allow differentiating between instances of the same PromptRequest.
+        // (Currently for some prompts it is possible to be requested and shown multiple times one on top of the other)
+        return super.hashCode() + System.currentTimeMillis().toInt()
+    }
```

Comment:
Haven't looked at the whole PR, but this part seems unusual. Is there a reason we cannot generate a UUID? This kind of breaks the `hashCode` behavior?

Reply:
Thank you!
Adding `currentTimeMillis` which is different on every `hashCode` call is not the right approach.
Will go with generating a UUID and use that to identify `PromptRequest`s and only override `equals`

Reply:
@Mugurell what happen if you do:

`sealed class PromptRequest(val uid : String = UUID.randomUUID().toString()) {`

Should other children classes will inherit the `uid` propperty and have it as part of the default implemented auto generated data class `equals` and `hashCode` functions?

Reply:
This is another alternative, but I'm not sure if it's worst than overriding/maintaining `equals` and `hashCode` 
<img width="1849" alt="image" src="https://user-images.githubusercontent.com/773158/119077962-46a65800-b9c3-11eb-86e0-c51a41aaa42f.png">

Reply:
Maybe this will more confusing than just overriding/maintaining `equals` and `hashCode`, I think it will be better to just keep it as it's, unfortunately data classes don't play well with inheritance  :(

Reply:
Thank you!
Tried the above suggestion and saw that `hashCode` and `equals` of subclasses will not also check `uid`.
And even with my first approach subclasses implementations of these methods will override super and not verify super's `uid` so if we need to support multiple PromptRequest instances with the same details every subclass will have to have their own `uid` property.